### ISSUE TITLE
If the command server is started in the background, the console becomes silent

### ISSUE THREAD

**Issue by thymol0** [radareorg/radare2#19118]
## Environment

```sh
Sun Sep 19 22:22:23 GMT 2021
radare2 5.4.1 26822 @ linux-x86-64 git.5.4.0-145-g719561644
commit: 719561644889df23b4a5a475963cfb008251fccf build: 2021-09-19__21:40:03
Linux x86_64
```

## Description

When a background task is executed the console doesn't print anything until the task is finished.

## Test

```
$ r2 -
[0x00000000]> & =t9000
[0x00000000]> Listening for commands on port 9000
[0x00000000]>
[0x00000000]> pd 1
```
No result will be printed to the console. If you break the background task then the queued console buffer will be printed.
```
[0x00000000]> &b 1
Task 1 finished
[0x00000000]> 
            0x00000000      0000           add byte [rax], al
```

**Comment by thymol0** [comment]
I am trying to fix this issue myself, my understanding so far:

1. There is only one global console with only one active context, this context holds the console buffer where the output is buffered before radare decides to print if the function called is `r_core_cmd` or to return it as a string if the function called is `r_core_cmd_str`.
2. Each background tasks runs in a dedicated thread.
3. Background tasks use `r_core_cmd_str`.
4. `r_core_cmd_str` disables console printing.

Since there is only one global console instance shared between all threads with only one active context, then all background tasks will be writing to the same context (i.e same buffer).

The solution is to either make an instance of the console for each thread or to make the console context different for each thread.

@trufae can you guide me here? running `grep -r -e '->context' libr| wc -l` return 269 matches, while I have no problem replacing every access to the console context with a a call to some function that returns the thread local context like `r_cons_get_context` or something, do you have something else in mind?

**Comment by trufae** [comment]
Thats how tasks work. The output of the task is saved in a buffer. Use the & command to collect the output

**Comment by trufae** [comment]
Oh sorry missreas the issue affects shell commands not the output of the task! Then thats an issue

**Comment by thymol0** [comment]
It seems the issue is quite deep. The `core` is shared and it is not meant to be used concurrently for example if two tasks run concurrently and one task changes the config to `e asm.hints=false` while the other sets `e asm.hints=true`, the result will be incorrect for one of them. In order to solve this systematically there will be quite a bit of restructuring in `libr/core/` to make it centered around tasks instead of being centered on the `core`.

It turns out I do not really need the entire radare core to become concurrent, I just need to run the command server in the background - while I am working with radare - to be accessed by a python script on the same machine. I will probably just go to the server command `=` and turn on console printing.

If someone needs radare to be concurrent, they can open another issue, I would be interested to know about it.

**Comment by trufae** [comment]
Tasks switch context when sleep begins and ends, so they are basically coroutines using mutex when r2 is waiting for user input.

so maybe the problem is related to an extra rconspush pop? And the output gets stacked in the wrong  buffer.

the cons buffer should be context specific and switch to it when a task runs. So thats a wrong behaviour.

**Comment by thymol0** [comment]
I have indeed discovered that when a task is waken up it calls `r_cons_context_load` and replaces the active console context with its own, but this only works if tasks do not run in parallel, in this case the command server runs in a dedicated thread and it never yields.

**Comment by thymol0** [comment]
I stumbled upon [this blog post](https://radareorg.github.io/blog/posts/background_tasks/) by thestr4ng3r which describes the tasks subsystem in radare and its limitations.

**Comment by trufae** [comment]
That is not a limitation for the task subsystem, it was working a while ago

**Comment by thymol0** [comment]
Thanks.
